### PR TITLE
Add option to automatically generate ros2_control's fake components description file

### PR DESCRIPTION
**Pull request by JafarAbdi**
### Description

Reduces the number of required trajectory execution config files for sim by 3 (out of 5), all the information needed to generate the fake components description file is already in the URDF file

See the discussion in  https://github.com/ros-planning/moveit2/issues/1261

### Checklist
- [ ] **Required by CI**: Code is auto formatted using [clang-format](http://moveit.ros.org/documentation/contributing/code)
- [ ] Extend the tutorials / documentation [reference](http://moveit.ros.org/documentation/contributing/)
- [ ] Document API changes relevant to the user in the [MIGRATION.md](https://github.com/ros-planning/moveit/blob/master/MIGRATION.md) notes
- [ ] Create tests, which fail without this PR [reference](https://ros-planning.github.io/moveit_tutorials/doc/tests/tests_tutorial.html)
- [ ] Include a screenshot if changing a GUI
- [ ] While waiting for someone to review your request, please help review [another open pull request](https://github.com/ros-planning/moveit/pulls) to support the maintainers

[//]: # "You can expect a response from a maintainer within 7 days. If you haven't heard anything by then, feel free to ping the thread. Thank you!"

### CODE REVIEW

File: moveit_configs_utils/moveit_configs_utils/moveit_configs_builder.py [hunk 1]
```diff
@@ -51,17 +51,63 @@
...
 from launch_param_builder import ParameterBuilder, load_yaml, load_xacro
-
+from jinja2 import Template
+from urdf_parser_py.urdf import URDF
+from xml.dom.minidom import Document, parseString
 
 moveit_configs_utils_path = Path(get_package_share_directory("moveit_configs_utils"))
 
 
+def generate_fake_system_description(
```
Comment (AndyZe) [inline moveit_configs_utils/moveit_configs_utils/moveit_configs_builder.py:68]: Suggest renaming this

```suggestion
def generate_fake_controllers(
```

File: moveit_configs_utils/moveit_configs_utils/moveit_configs_builder.py [hunk 2]
```diff
@@ -51,17 +51,63 @@
...
+            <param name="multiplier">{{joint.mimic.multiplier}}</param>
+            {%- endif %}
+            {%- endif %}
+            <command_interface name="position"/>
+            <state_interface name="position">
+              <param name="initial_value">{{initial_position.get(joint.name, 0.0)}}</param>
+            </state_interface>
+            <state_interface name="velocity">
+              <param name="initial_value">0.0</param>
+            </state_interface>
```
Comment (AndyZe) [inline moveit_configs_utils/moveit_configs_utils/moveit_configs_builder.py:95]: This defaults to 0.

```suggestion
            <state_interface name="velocity"/>
```

File: moveit_configs_utils/moveit_configs_utils/moveit_configs_builder.py [hunk 3]
```diff
@@ -288,21 +334,61 @@ def __init__(
 
         self.__robot_description = robot_description
 
-    def robot_description(self, file_path: Optional[str] = None, mappings: dict = None):
+    def robot_description(
+        self,
+        file_path: Optional[str] = None,
+        mappings: dict = None,
+        auto_generate_fake_components: bool = True,
```
Comment (AndyZe) [inline moveit_configs_utils/moveit_configs_utils/moveit_configs_builder.py:341]:
```suggestion
        auto_generate_fake_controllers: bool = True,
```

File: moveit_configs_utils/moveit_configs_utils/moveit_configs_builder.py [hunk 4]
```diff
@@ -288,21 +334,61 @@ def __init__(
...
+        initial_positions: Union[str, dict] = None,
+    ):
         """Load robot description.
 
         :param file_path: Absolute or relative path to the URDF file (w.r.t. robot_name_moveit_config).
-        :param mappings: mappings to be passed when loading the xacro file.
+        :param mappings: Mappings to be passed when loading the xacro file.
+        :param auto_generate_fake_components: If true moveit_configs_utils will automatically generate the ros2_control description for the fake components
```
Comment (AndyZe) [inline moveit_configs_utils/moveit_configs_utils/moveit_configs_builder.py:348]:
```suggestion
        :param auto_generate_fake_controllers: If true moveit_configs_utils will automatically generate the ros2_control description for the fake components
```

File: moveit_configs_utils/moveit_configs_utils/moveit_configs_builder.py [hunk 5]
```diff
@@ -288,21 +334,61 @@ def __init__(
...
+        :param initial_positions: A dictionary of joint_name -> initial_joint_position, only used when ros2_control description is automatically generated
```
Comment (AndyZe) [inline moveit_configs_utils/moveit_configs_utils/moveit_configs_builder.py:349]:
```suggestion
        :param initial_positions: A dictionary of joint_name -> initial_joint_position
```

File: moveit_configs_utils/moveit_configs_utils/moveit_configs_builder.py [hunk 6]
```diff
@@ -288,21 +334,61 @@ def __init__(
...
+                raise RuntimeError(
+                    f"Can't auto-generate ros2_control fake components description, provided robot_description file '{robot_description_file_path}' already contains 'ros2_control' tag"
+                    "Make sure to set auto_generate_fake_components to false when calling robot_description(..., auto_generate_fake_components = false)"
+                )
+            if type(initial_positions) == str:
+                initial_positions = load_yaml(self._package_path / initial_positions)[
+                    "initial_positions"
+                ]
+            if initial_positions is None:
+                # For backward compatibility check for config/initial_positions.yaml
```
Comment (AndyZe) [inline moveit_configs_utils/moveit_configs_utils/moveit_configs_builder.py:377]: I think you should just print a warning here. No need to maintain backward compatability in my opinion

File: moveit_configs_utils/moveit_configs_utils/moveit_configs_builder.py [hunk 7]
```diff
@@ -288,21 +334,61 @@ def __init__(
...
+                initial_positions_file = (
+                    self._package_path / "config" / "initial_positions.yaml"
+                )
+                if initial_positions_file.exists():
+                    initial_positions = load_yaml(initial_positions_file)[
+                        "initial_positions"
+                    ]
+            ros2_control_fake_components_desc = generate_fake_system_description(
```
Comment (AndyZe) [inline moveit_configs_utils/moveit_configs_utils/moveit_configs_builder.py:385]:
```suggestion
            ros2_control_fake_components_desc = generate_fake_controllers(
```

File: moveit_configs_utils/test/test_moveit_resources_configs.py [hunk 8]
```diff
@@ -14,10 +17,185 @@ def test_panda_with_gripper_config():
...
-            .robot_description(file_path="config/panda.urdf.xacro")
             .robot_description_semantic(file_path="config/panda.srdf")
             .trajectory_execution(file_path="config/gripper_moveit_controllers.yaml")
         )
         assert moveit_config.to_dict()
     except RuntimeError as e:
         assert False, f"moveit_resources_panda gripper config failed to build: {e}"
+
+
+def test_generated_panda_fake_components():
```
Comment (AndyZe) [inline moveit_configs_utils/test/test_moveit_resources_configs.py:28]:
```suggestion
def test_generated_panda_fake_controllers():
```

**Review by AndyZe — APPROVED** [review]
I'm OK with this. It looks like you satisfied my concern from #1261

**Comment by DLu** [comment]
As mentioned in the maintainers meeting last week, I have some code that does this the MSA side. Currently both `fanuc` and `panda` have it as part of their moveit configs. 

I guess my main question is: when is the [ros2_control tag](https://github.com/ros-planning/moveit_resources/blob/ros2/fanuc_moveit_config/config/fanuc.ros2_control.xacro#L7) need to be used?

**Comment by DLu** [comment]
> Any time you're using ros2_control at all, that tag has to be in the URDF somewhere. Simulation or hardware

What is the "fake" part then?

**Comment by AndyZe** [comment]
> What is the "fake" part then?

Sorry I didn't understand the question completely at first. For you auto-generating these templates, this should always be fine. It's what allows the RViz simulation without needing Gazebo/Ignition.

```
            <hardware>
                <plugin>fake_components/GenericSystem</plugin>
            </hardware>
```

If people want to start using real hardware, they'll need to add their own tag. Like [here](https://github.com/UniversalRobots/Universal_Robots_ROS2_Description/blob/ros2/urdf/ur.ros2_control.xacro) for the UR driver:

```
        <xacro:if value="${sim_gazebo}">
          <plugin>gazebo_ros2_control/GazeboSystem</plugin>
        </xacro:if>
        <xacro:if value="${sim_ignition}">
          <plugin>ign_ros2_control/IgnitionSystem</plugin>
        </xacro:if>
        <xacro:if value="${use_fake_hardware}">
          <plugin>fake_components/GenericSystem</plugin>
          <param name="fake_sensor_commands">${fake_sensor_commands}</param>
          <param name="state_following_offset">0.0</param>
        </xacro:if>
        <xacro:unless value="${use_fake_hardware or sim_gazebo or sim_ignition}">
          <plugin>ur_robot_driver/URPositionHardwareInterface</plugin>
```

**Comment by DLu** [comment]
@JafarAbdi Do you think that https://github.com/ros-planning/moveit2/pull/1299 covers everything you wanted to do with this PR?

**Comment by tylerjw** [comment]
@AndyZe or @JafarAbdi do you know what should be done next with this change?

**Comment by tylerjw** [comment]
This PR has many conflicts and was abandoned. It seems like a useful change so I'll create an issue to track finishing this work.

**Pull request closed without merging**